test/utils.test.js

```diff
@@ -14,37 +38,367 @@ describe('utils', function() {
+        const child = ['p'];
+        utils.appendChild(jml, child);
+        assert.deepEqual(jml, ['div', child]);
+      });
+
+      it('should append element with all descendents', function() {
+        const jml = ['div', { a: 1}];
+        const subtree = ['section',
+          ['p', 'hello'],
+          ['p', 'world'],
+        ];
+        utils.appendChild(jml, subtree);
+        assert.deepEqual(jml, ['div', { a: 1 }, subtree]);
+      });
+
+      it('should execute an optional callback before appending (for JBST use)', function(done) {
+        const jml = ['div'];
+        const el = ['p'];
+        utils.appendChild(jml, el, function onAppend(parent, child) {
+          assert.strictEqual(jml, parent);
+          assert.strictEqual(el, child);
+          assert.deepEqual(jml, ['div']);
+          done();
+        });
+      });
+    });
+
+    describe('when appending a raw child', function() {
+      // @TODO - this will currently throw an error due to missing function `isRaw`
+      xit('should append to the parent', function() {
+        const jml = ['div'];
+        const child = new function RawChild() {};
+        utils.appendChild(jml, child);
+        assert.strictEqual(jml, ['div', child]);
+      });
+    });
+
+    describe('when appending attributes', function() {
+      // @TODO - this will currently throw an error due to missing function `isRaw`
+      xit('should add attributes to the parent element', function() {
+        const jml = ['div'];
+        const attrs = { a: 1 };
+        utils.appendChild(jml, attrs);
+        assert.deepEqual(jml, ['div', { a: 1 }]);
+      });
+    });
+
+    describe('when appending a defined, non-null value', function() {
+      it('should append value to existing text node', function() {
+        const jml = ['p', { a: 1 }, 'foo'];
+        utils.appendChild(jml, 'bar');
+        assert.deepEqual(jml, ['p', { a: 1 }, 'foobar']);
+      });
+
+      it('should coerce value to a string', function() {
+        const jml = ['p', { a: 1 }, 'foo'];
+        utils.appendChild(jml, 5);
+        utils.appendChild(jml, true);
+        assert.deepEqual(jml, ['p', { a: 1 }, 'foo5true']);
+      });
+
+      it('should append text node', function() {
+        const jml = ['p'];
+        utils.appendChild(jml, 'hello');
+        assert.deepEqual(jml, ['p', 'hello']);
+      });
+
+      it('should not append an empty string', function() {
+        const jml = ['p'];
+        utils.appendChild(jml, '');
+        assert.deepEqual(jml, ['p']);
+      });
+
+      // What is the use case for this behavior?
+      // @TODO investigate and document
+      it('should append child string to empty parent array', function() {
+        const jml = [];
+        utils.appendChild(jml, '');
+        assert.deepEqual(jml, ['']);
+
+        utils.appendChild(jml, 'x');
+        assert.deepEqual(jml, ['', 'x']);
+
+        utils.appendChild(jml, 'z');
+        assert.deepEqual(jml, ['', 'xz']);
+      });
```

Comment:
I added this test based on the condition to allow a [parent element with no length](https://github.com/benjycui/jsonml.js/blob/master/lib/utils.js#L132). I'm not really sure what the purpose is. To build a fragment?

Reply:
Just raise a new issue, I need to to read the source code, and may be re-design these APIs.

Reply:
Filed issue: #11